design-proposals/klip-9-table-functions.md

```diff
@@ -0,0 +1,206 @@
+It's a function that returns a set of zero or more rows. Contrast this to a scalar function which
+returns a single value. We will call them _table functions_ in the remainder of this document.
+
+Table functions are typically used in FROM clauses with traditional databases. Some databases
+such as PostgreSQL and Hive also allow table functions to appear in SELECT clauses.
+
+An example of a table function from Apache Hive is `explode` - this takes a non scalar type such as
+an ARRAY or MAP and explodes it to zero or more rows.
+
+`select explode(array('A','B','C'));`
+
+Would return three rows:
+
+```
+A
+B
+C
+```
+
+Table functions can be thought of as analagous to the `FlatMap` operation commonly found in 
+functional programming or stream processing frameworks such as Kafka Streams.
+
+Please see the following links for further background information:
+
+* [Apache Hive Table Generating Functions](https://cwiki.apache.org/confluence/display/Hive/LanguageManual+UDF#LanguageManualUDF-Built-inTable-GeneratingFunctions(UDTF)
+* [Microsoft SQL Server table functions](https://docs.microsoft.com/en-us/dotnet/framework/data/adonet/sql/linq/how-to-use-table-valued-user-defined-functions)
+* [Google BigQuery (see UNNEST)](https://cloud.google.com/bigquery/docs/reference/standard-sql/query-syntax)
+* [PostgreSQL (see UNNEST)](http://www.postgresql.cn/docs/9.5/functions-array.html) 
+* [Kafka Streams (see FlatMap)](https://kafka.apache.org/20/documentation/streams/developer-guide/dsl-api.html)
+
+## Motivation
+
+We propose to support table functions in SELECT clauses in KSQL.
+
+There is [demand from KSQL users](https://github.com/confluentinc/ksql/issues/527) for a way to explode non scalar types such
+as ARRAY or MAP from columns in KSQL streams or tables into multiple rows. This is the main
+motivation for proposing this feature.
+
+Table funtions would also be useful for other cases such as splitting a string into words, or any
+case where you need to go from a single row to zero or more, for example, unaggregating data or
+duplicating rows.
+
+Here's an example of exploding an array from a column into multiple rows:
+
+We have a stream containing a batch of sensor readings from a particular sensor. For further
+stream processing we want to split them out into separate messages:
+
+Given:
+
+```
+{sensor_id:12345 readings: [23, 56, 3, 76, 75]}
+```
+
+We want to get:
+
+```
+{sensor_id:12345 reading: 23}
+{sensor_id:12345 reading: 56}
+{sensor_id:12345 reading: 3}
+{sensor_id:12345 reading: 76}
+{sensor_id:12345 reading: 75}
+```
+
+This could be accomplished by the following CSAS:
+
+```
+CREATE STREAM exploded_stream AS
+    SELECT sensor_id, EXPLODE(readings) AS reading FROM batched_readings;
+```
+
+Note that any scalar values in the select will be duplicated for each row in the output.
+
+
+## What is in scope
+
+We would like to implement
+
+* Table functions on the SELECT clause only
+* Multiple table functions on the SELECT clause
```

Comment:
What would be the semantics for 2+ table functions? Would it work like a zipmap function (e.g. zip a new row per index in each collection)?

Reply:
Yep, exactly like a zipmap. There's a bit of detail about this in the design section.

Reply:
Ah, there it is. I foolishly skipped the design section while reading. :)

Reply:
The other option here is to return a cross product when there are multiple table valued functions. 

So if you had functions `func1(col1)` and `func2(col2)` returning rows `[x, y]` and `[a, b]` respectively, you could output:

```
x, a
x, b
y, a
y, b
```

The current proposal would return

```
x, a
y, b
```

Did we consider supporting both semantics? Is there any value in doing so?

Reply:
At least Hive and PostgreSQL support UDTFs on SELECT clause, and this seems the simplest way to express a zipjoin across multiple UDTFs on the select. A zipjoin seems the most useful use case.
Expressing a zipjoin in SQL on a JOIN is very non-unintuitive IMHO (it requires correlated subqueries). I think there's a comment and illustration to that effect on the original discussion.
I don't the cartesian product is terribly useful compared to a zipjoin.